app/views/admin/api_keys.py

```diff
@@ -0,0 +1,225 @@
+
+    def get(self, request, *args, **kwargs):
+        """Serves get requests.
+
+        Args:
+            request: Unused.
+            *args: Unused.
+            **kwargs: Unused.
+
+        Returns:
+            HttpResponse: A HTTP response with the API key list.
+        """
+        del request, args, kwargs  # unused
+        auths = model.Authorization.all().filter(
+            'repo = ', self.env.repo or '*')
+        return self.render(
+            'admin_api_keys_list.html',
+            admin_api_keys_url=self.build_absolute_path(
+                '/admin/api_keys', self.env.repo or 'global'),
+            user=self.env.user,
+            user_email=self.env.user.email(),
+            authorizations=auths,
+            xsrf_token=self.xsrf_tool.generate_token(
+               self.env.user.user_id(), 'admin_api_keys'))
+
+
+class ApiKeyManagementView(views.admin.base.AdminBaseView):
+    """The API key management view, for creating or editing keys."""
+
+    ACTION_ID = 'admin/api_keys'
+
+    def get_params(self):
+        return views.base.read_params(
+            super(ApiKeyManagementView, self).get_params(),
+            self.request,
+            get_params={
+                'log_key': utils.strip,
+            },
+            post_params={
+                'authorization_key': utils.strip,
+                'believed_dead_permission': utils.validate_checkbox_as_bool,
+                'contact_name': utils.strip,
+                'contact_email': utils.strip,
+                'domain_write_permission': utils.strip,
+                'edit_form': utils.strip,
+                'full_read_permission': utils.validate_checkbox_as_bool,
+                'is_valid': utils.validate_checkbox_as_bool,
+                'key': utils.strip,
+                'mark_notes_reviewed': utils.validate_checkbox_as_bool,
+                'organization_name': utils.strip,
+                'read_permission': utils.validate_checkbox_as_bool,
+                'search_permission': utils.validate_checkbox_as_bool,
+                'stats_permission': utils.validate_checkbox_as_bool,
+                'subscribe_permission': utils.validate_checkbox_as_bool,
+            })
+
+    def get(self, request, *args, **kwargs):
+        if self.params.get('log_key'):
+            management_log_key = self.params.log_key
+            management_log = db.get(management_log_key)
+            message = ''
+            if management_log.action == model.ApiKeyManagementLog.CREATE:
+                message = t.ugettext(
+                    'A new API key has been created successfully.')
+            elif management_log.action == model.ApiKeyManagementLog.UPDATE:
+                message = t.ugettext(
+                    'The API key has been updated successfully.')
+            return self._render_form(management_log.authorization, message)
+        else:
+            return self._render_form()
+
+    def post(self, request, *args, **kwargs):
```

Comment:
this method is pretty long. would it be possible to break it up?

Reply:
A little? I factored out the Authorization-creating code and added a general XSRF enforcement function to the base admin handler (it works for 403s because we don't include any message with those, but I can't do the same with cases that involve a custom error message).

Reply:
:+1: